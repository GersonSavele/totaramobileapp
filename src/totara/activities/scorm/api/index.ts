/**
 * This file is part of Totara Mobile
 *
 * Copyright (C) 2019 onwards Totara Learning Solutions LTD
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 * @author Tharaka Dushmantha <tharaka.dushmantha@totaralearning.com
 */

import gql from "graphql-tag";
import { config } from "@totara/lib";

const scormQuery = gql`
  query totara_mobile_scorm($scormid: core_id!) {
    scorm: mod_scorm_scorm(scormid: $scormid) {
      id
      courseid
      name
      scormtype
      reference
      intro(format: PLAIN)
      version
      maxgrade
      grademethod
      whatgrade
      maxattempt
      forcecompleted
      forcenewattempt
      lastattemptlock
      masteryoverride
      displaycoursestructure
      skipview
      nav
      navpositionleft
      navpositiontop
      auto
      width
      height
      timeopen
      timeclose
      displayactivityname
      autocommit
      allowmobileoffline
      completion
      completionview
      completionstatusrequired
      completionscorerequired
      completionstatusallscos
      packageUrl: package_url
      launchUrl: launch_url
      repeatUrl: repeat_url
      attemptsCurrent: attempts_current
      calculatedGrade: calculated_grade
      offlinePackageUrl: offline_package_url
      offlinePackageContentHash: offline_package_contenthash
      offlinePackageScoIdentifiers: offline_package_sco_identifiers
      attempts {
        attempt
        timestarted
        gradereported
      }
      # Deprecated properties
      type: scormtype
      description: intro(format: PLAIN)
      attemptsMax: maxattempt
      attemptsCurrent: attempts_current
      attemptsForceNew: forcenewattempt
      attemptsLockFinal: lastattemptlock
      autoContinue: auto
      offlineAttemptsAllowed: allowmobileoffline
      __typename
    }
  }
`;

/**
 * returns an map object with all scorms
 */
const scormActivitiesRecordsQuery = gql`
  query get_scorm_bundle {
    scormBundles @client
  }
`;

const scormFeedbackQuery = gql`
  query totara_mobile_scorm_current_status($scormid: core_id!) {
    status: mod_scorm_current_status(scormid: $scormid) {
      maxattempt
      attempts_current
      completion
      completionview
      completionstatusrequired
      completionscorerequired
      completionstatusallscos
      completionstatus
      gradefinal
      grademax
      gradepercentage
      __typename
    }
  }
`;

const mutationAttempts = gql`
  mutation mod_scorm_save_offline_attempts(
    $scormid: core_id!
    $attempts: [mod_scorm_attempt!]!
  ) {
    attempts: mod_scorm_save_offline_attempts(
      scormid: $scormid
      attempts: $attempts
    ) {
      attempts_accepted
      maxattempt
      attempts_current
      completion
      completionview
      completionstatusrequired
      completionscorerequired
      completionstatusallscos
      completionstatus
      gradefinal
      grademax
      gradepercentage
      __typename
    }
  }
`;

const fetchLastAttemptResult = ({
  scormId,
  apiKey,
  host
}: {
  scormId: string;
  apiKey: string;
  host: string;
}): Promise<Response> => {
  // fetch from global
  // eslint-disable-next-line no-undef
  return fetch(config.apiUri(host), {
    method: "POST",
    body: JSON.stringify({
      operationName: "totara_mobile_scorm_current_status",
      variables: { scormid: scormId }
    }),
    headers: {
      Accept: "application/json",
      Authorization: `Bearer ${apiKey}`
    }
  }).then((response) => {
    if (response.status === 200) {
      return response.json();
    } else {
      throw new Error(response.status.toString());
    }
  });
};

export {
  scormQuery,
  scormFeedbackQuery,
  scormActivitiesRecordsQuery,
  mutationAttempts,
  fetchLastAttemptResult
};
