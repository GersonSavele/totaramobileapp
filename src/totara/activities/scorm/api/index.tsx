/**
 * This file is part of Totara Mobile
 *
 * Copyright (C) 2019 onwards Totara Learning Solutions LTD
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 * @author Tharaka Dushmantha <tharaka.dushmantha@totaralearning.com
 */

import React from "react";
import gql from "graphql-tag";
import { ScormActivity } from "@totara/types";
import { Query } from "react-apollo";
import console = require("console");
import { View } from "native-base";


class ScormQuery extends Query<Response, Variables> {}

type Response = {
    scorm: ScormActivity;
  }
  
interface Variables {
    id: number;
  }

const ScormGQLQuery = gql` 
query scorm($id: ID!) {
    scorm(id: $id) {
      id
      webEntryUrl
      currentAttempt
      maxAttempt
      score
      isAvailable
    }
  }
`;

const GetScormQuery = ({props,id}: any) => {
   <ScormQuery  query= { ScormGQLQuery } variables = {{ id : id }}>
    {props}
   </ScormQuery>
}



export default GetScormQuery;
