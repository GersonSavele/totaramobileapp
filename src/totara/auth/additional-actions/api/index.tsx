/**
 * This file is part of Totara Enterprise.
 *
 * Copyright (C) 2019 onwards Totara Learning Solutions LTD
 *
 * Totara Enterprise is provided only to Totara Learning Solutions
 * LTDâ€™s customers and partners, pursuant to the terms and
 * conditions of a separate agreement with Totara Learning
 * Solutions LTD or its affiliate.
 *
 * If you do not have an agreement with Totara Learning Solutions
 * LTD, you may not access, use, modify, or distribute this software.
 * Please contact [sales@totaralearning.com] for more information.
 */

import * as React from "react";
import { Query } from "react-apollo";
import gql from "graphql-tag";
import { Me } from "@totara/types";

type Response = {
  me: Me;
};

export type QueryResult = {
  loading: boolean;
  data: Response;
  error: Error;
  refetch: () => void;
};

const QueryMe = gql`
  query totara_mobile_me {
    me: totara_mobile_me {
      user {
        id
        firstname
        lastname
        fullname
        email
      }
      system {
        wwwroot
        apiurl
        release
        request_policy_agreement
        request_user_consent
        request_user_fields
        password_change_required
      }
    }
  }
`;

const GetMe = ({ props }: any) => {
  return (
    <Query<Response> query={QueryMe}>
      {({ loading, data, error, refetch }) =>
        props({
          loading: loading,
          data: data,
          error: error,
          refetch: refetch
        })
      }
    </Query>
  );
};

export { GetMe, QueryMe };
