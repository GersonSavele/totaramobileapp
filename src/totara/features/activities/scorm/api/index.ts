/**
 * This file is part of Totara Enterprise.
 *
 * Copyright (C) 2020 onwards Totara Learning Solutions LTD
 *
 * Totara Enterprise is provided only to Totara Learning Solutions
 * LTDâ€™s customers and partners, pursuant to the terms and
 * conditions of a separate agreement with Totara Learning
 * Solutions LTD or its affiliate.
 *
 * If you do not have an agreement with Totara Learning Solutions
 * LTD, you may not access, use, modify, or distribute this software.
 * Please contact [sales@totaralearning.com] for more information.
 */

import { gql } from "@apollo/client";
import { config } from "@totara/lib";
import { AUTH_HEADER_FIELD } from "@totara/lib/constants";

//TODO: rename - use alias - when backend finishes TL-26268
const scormQuery = gql`
  query totara_mobile_scorm($scormid: core_id!) {
    scorm: mod_scorm_scorm(scormid: $scormid) {
      id
      courseid
      name
      scormtype
      reference
      intro(format: PLAIN)
      version
      maxgrade
      maxattempt
      forcecompleted
      forcenewattempt
      lastattemptlock
      masteryoverride
      displaycoursestructure
      skipview
      nav
      navpositionleft
      navpositiontop
      auto
      width
      height
      timeclose
      displayactivityname
      autocommit
      allowmobileoffline
      completion
      completionview
      completionstatusrequired
      completionstatusallscos
      packageUrl: package_url
      launchUrl: launch_url
      repeatUrl: repeat_url
      attemptsCurrent: attempts_current
      calculatedGrade: calculated_grade
      offlinePackageUrl: offline_package_url
      offlinePackageContentHash: offline_package_contenthash
      offlinePackageScoIdentifiers: offline_package_sco_identifiers
      attempts {
        attempt
        timestarted
        gradereported
      }
      # Deprecated properties
      type: scormtype
      description: intro(format: PLAIN)
      attemptsForceNew: forcenewattempt
      attemptsLockFinal: lastattemptlock
      autoContinue: auto
      offlineAttemptsAllowed: allowmobileoffline
      newAttemptDefaults: attempt_defaults
      whatgrade
      grademethod
      timeopen
      completionscorerequired
      attemptsMax: maxattempt
      showGrades: showgrades
      __typename
    }
  }
`;

/**
 * returns an map object with all scorms
 */
const scormActivitiesRecordsQuery = gql`
  query get_scorm_bundle {
    scormBundles @client
  }
`;

const scormFeedbackQuery = gql`
  query totara_mobile_scorm_current_status($scormid: core_id!) {
    status: mod_scorm_current_status(scormid: $scormid) {
      maxattempt
      attempts_current
      completion
      completionview
      completionstatusrequired
      completionscorerequired
      completionstatusallscos
      completionstatus
      gradefinal
      grademax
      gradepercentage
      __typename
    }
  }
`;

const mutationAttempts = gql`
  mutation mod_scorm_save_offline_attempts($scormid: core_id!, $attempts: [mod_scorm_attempt!]!) {
    attempts: mod_scorm_save_offline_attempts(scormid: $scormid, attempts: $attempts) {
      attempts_accepted
      maxattempt
      attempts_current
      completion
      completionview
      completionstatusrequired
      completionscorerequired
      completionstatusallscos
      completionstatus
      gradefinal
      grademax
      gradepercentage
      __typename
    }
  }
`;

const fetchLastAttemptResult = ({
  scormId,
  apiKey,
  host
}: {
  scormId: string;
  apiKey: string;
  host: string;
}): Promise<Response> => {
  // fetch from global
  // eslint-disable-next-line no-undef
  return fetch(config.apiUri(host), {
    method: "POST",
    body: JSON.stringify({
      operationName: "totara_mobile_scorm_current_status",
      variables: { scormid: scormId }
    }),
    headers: {
      Accept: "application/json",
      [AUTH_HEADER_FIELD]: apiKey
    }
  }).then((response) => {
    if (response.status === 200) {
      return response.json();
    } else {
      throw new Error(response.status.toString());
    }
  });
};

export { scormQuery, scormFeedbackQuery, scormActivitiesRecordsQuery, mutationAttempts, fetchLastAttemptResult };
