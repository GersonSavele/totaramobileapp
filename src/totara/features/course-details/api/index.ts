/**
 * This file is part of Totara Mobile
 *
 * Copyright (C) 2019 onwards Totara Learning Solutions LTD
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 * @author Jun Yamog <jun.yamog@totaralearning.com>
 *
 */

import { graphql } from "react-apollo";
import gql from "graphql-tag";
import { Course } from "@totara/types";
import { NavigationInjectedProps } from "react-navigation";

const query = gql`
  query totara_mobile_course($courseid: ID!) {
      course: core_course(courseid: $courseid) {
          id
          fullname
          shortname
          summary(format: PLAIN)
          startdate(format: ISO8601)
          enddate(format: ISO8601)
          lang
          imageSrc: image
          sections {
              id
              title
              available
              availablereason(format: PLAIN)
              data: modules {
                  id
                  modtype
                  name
                  available
                  availablereason(format: PLAIN)
                  viewurl
                  completion
                  completionstatus
                  showdescription
                  description(format: PLAIN)
                  gradefinal
                  gradepercentage
              }
          }
          completion {
              statuskey
              progress
              timecompleted(format: ISO8601)
          }
      }
  }
`;

type CourseId = {
  courseid: number;
};

export type Response = {
  course: Course;
} & NavigationInjectedProps<CourseId>;

export const getCourse = graphql<NavigationInjectedProps, Response>(query, {
  options: props => ({
    variables: { courseid: props.navigation.state.params.courseId }
  }),
  props: ({ data }) => ({ ...data })
  // needed magic for double HOC, I have no idea why extracting data here makes wrapped totara.components props passed down.  As opposed to extracting data on the calling function
  // it is just stated to be better with static typing, but it doesn't work w/o this.  see: https://www.apollographql.com/docs/react/recipes/static-typing.html#props
});
