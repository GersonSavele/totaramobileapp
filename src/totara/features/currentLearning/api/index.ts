/**
 * This file is part of Totara Enterprise.
 *
 * Copyright (C) 2019 onwards Totara Learning Solutions LTD
 *
 * Totara Enterprise is provided only to Totara Learning Solutions
 * LTDâ€™s customers and partners, pursuant to the terms and
 * conditions of a separate agreement with Totara Learning
 * Solutions LTD or its affiliate.
 *
 * If you do not have an agreement with Totara Learning Solutions
 * LTD, you may not access, use, modify, or distribute this software.
 * Please contact [sales@totaralearning.com] for more information.
 */

import { gql } from "@apollo/client";
import { config } from "@totara/lib";

const query = config.mobileApi.persistentQuery
  ? gql`
      query totara_mobile_current_learning {
        currentLearning: totara_mobile_current_learning {
          id
          itemtype
          itemcomponent
          shortname
          fullname
          summary: description(format: PLAIN)
          progress
          urlView: url_view
          duedate(format: ISO8601)
          duedateState: duedate_state
          native: mobile_coursecompat
          imageSrc: mobile_image
          summaryFormat: description_format
          __typename
        }
      }
    `
  : gql`
      query mobile_currentlearning_my_items {
        currentLearning: mobile_currentlearning_my_items {
          id
          itemtype
          itemcomponent
          shortname
          fullname
          summary: description(format: PLAIN)
          progress
          urlView: url_view
          duedate(format: ISO8601)
          duedateState: duedate_state
          native: mobile_coursecompat
          imageSrc: mobile_image
          summaryFormat: description_format
          __typename
        }
      }
    `;

export default query;
