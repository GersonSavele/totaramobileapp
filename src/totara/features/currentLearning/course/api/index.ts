/**
 * This file is part of Totara Enterprise.
 *
 * Copyright (C) 2020 onwards Totara Learning Solutions LTD
 *
 * Totara Enterprise is provided only to Totara Learning Solutions
 * LTDâ€™s customers and partners, pursuant to the terms and
 * conditions of a separate agreement with Totara Learning
 * Solutions LTD or its affiliate.
 *
 * If you do not have an agreement with Totara Learning Solutions
 * LTD, you may not access, use, modify, or distribute this software.
 * Please contact [sales@totaralearning.com] for more information.
 */

import { gql } from "@apollo/client";
import { config } from "@totara/lib";
import { AUTH_HEADER_FIELD } from "@totara/lib/constants";

const coreCourse = gql`
  query totara_mobile_course($courseid: core_id!, $guestpw: String) {
    mobile_course: totara_mobile_course(courseid: $courseid, guestpw: $guestpw) {
      course(courseid: $courseid) {
        id
        format
        fullname
        shortname
        summary(format: PLAIN)
        summaryformat
        startdate(format: ISO8601)
        enddate(format: ISO8601)
        lang
        image
        sections {
          id
          title
          available
          availablereason(format: PLAIN)
          summary(format: PLAIN)
          summaryformat
          data: modules {
            id
            instanceid
            modtype
            name
            available
            availablereason(format: PLAIN)
            viewurl
            completion
            completionstatus
            showdescription
            description(format: PLAIN)
            gradefinal
            gradepercentage
            descriptionformat
            __typename
          }
          __typename
        }
        criteriaaggregation
        criteria {
          id
          type
          typeaggregation
          criteria
          requirement
          status
          complete
          completiondate(format: ISO8601)
          __typename
        }
        showGrades: showgrades
        completionEnabled: completionenabled
        completion {
          id
          statuskey
          progress
          timecompleted(format: ISO8601)
          __typename
        }
        __typename
      }
      native: mobile_coursecompat
      imageSrc: mobile_image
      gradeFinal: formatted_gradefinal
      gradeMax: formatted_grademax
      __typename
    }
  }
`;

const courseSelfComplete = gql`
  mutation totara_mobile_completion_course_self_complete($courseid: core_id!) {
    core_completion_course_self_complete(courseid: $courseid)
  }
`;

const activitySelfComplete = gql`
  mutation totara_mobile_completion_activity_self_complete($cmid: core_id!, $complete: Boolean!) {
    core_completion_activity_self_complete(cmid: $cmid, complete: $complete)
  }
`;

type FetchParam = {
  instanceId: number;
  apiKey: string;
  host: string;
  modtype?: string;
};

const XMLHttpRequestPostFetch = (host, body, apiKey) => {
  return new Promise(function (resolve, reject) {
    // We decided to use XMLHttpRequest instead of fetch because fetch seems to have an issue being blocked by some missing certificate
    // please check https://github.com/facebook/react-native/issues/32931 and MOB-1144
    // eslint-disable-next-line no-undef
    var request = new XMLHttpRequest();
    request.open("POST", config.apiUri(host));
    request.setRequestHeader("Content-Type", "application/json;charset=UTF-8");
    request.setRequestHeader(AUTH_HEADER_FIELD, apiKey);

    request.onload = function () {
      if (request.status >= 200 && request.status < 300) {
        resolve(request.response);
      } else {
        reject({
          status: request.status,
          statusText: request.statusText
        });
      }
    };

    request.onerror = function () {
      reject({
        status: request.status,
        statusText: request.statusText
      });
    };

    request.send(body);
  });
};

const fetchResource = ({ instanceId, apiKey, host }: FetchParam): Promise<unknown> => {
  return XMLHttpRequestPostFetch(
    host,
    JSON.stringify({
      operationName: "totara_mobile_resource",
      variables: { resourceid: instanceId }
    }),
    apiKey
  );
};

const updateStateViewResource = ({ instanceId, modtype, apiKey, host }: FetchParam): Promise<unknown> => {
  return XMLHttpRequestPostFetch(
    host,
    JSON.stringify({
      operationName: "totara_mobile_completion_activity_view",
      variables: { cmid: instanceId, activity: modtype }
    }),
    apiKey
  );
};

export { coreCourse, courseSelfComplete, activitySelfComplete, fetchResource, updateStateViewResource };
