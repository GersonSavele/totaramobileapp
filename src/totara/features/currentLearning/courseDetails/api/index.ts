/**
 * This file is part of Totara Mobile
 *
 * Copyright (C) 2019 onwards Totara Learning Solutions LTD
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 * @author Jun Yamog <jun.yamog@totaralearning.com>
 *
 */

import gql from "graphql-tag";

const coreCourse = gql`
  query totara_mobile_course($courseid: ID!) {
    mobile_course: totara_mobile_course(courseid: $courseid) {
      course(courseid: $courseid) {
        id
        fullname
        shortname
        summary(format: PLAIN)
        startdate(format: ISO8601)
        enddate(format: ISO8601)
        lang
        image
        sections {
          id
          title
          available
          availablereason(format: PLAIN)
          data: modules {
            id
            instanceid
            modtype
            name
            available
            availablereason(format: PLAIN)
            viewurl
            completion
            completionstatus
            showdescription
            description(format: PLAIN)
            gradefinal
            gradepercentage
            __typename
          }
          __typename
        }
        criteriaaggregation
        criteria {
          id
          type
          typeaggregation
          criteria
          requirement
          status
          complete
          completiondate(format: ISO8601)
          __typename
        }
        completion {
          id
          statuskey
          progress
          timecompleted(format: ISO8601)
          gradefinal
          grademax
          __typename
        }
        __typename
      }
      native: mobile_coursecompat
      imageSrc: mobile_image
      __typename
    }
  }
`;

const courseSelfComplete = gql`
  mutation totara_mobile_completion_course_self_complete($courseid: core_id!) {
    core_completion_course_self_complete(courseid: $courseid)
  }
`;

export { coreCourse, courseSelfComplete };
