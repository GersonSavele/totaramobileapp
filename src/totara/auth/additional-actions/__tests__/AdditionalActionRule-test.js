/**
 * This file is part of Totara Mobile
 *
 * Copyright (C) 2019 onwards Totara Learning Solutions LTD
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 * @author Tharaka Dushmantha <tharaka.dushmantha@totaralearning.com>
 */

import { MockedProvider } from "@apollo/react-testing";
import renderer from "react-test-renderer";
import React from "react";
import wait from "waait";
import { Text } from "react-native";

import { QueryMe } from "../api";
import {
  AdditionalActionRule,
  AdditionalActionRuleCondition
} from "../AdditionalActionRule";
import { act } from "@testing-library/react-native";

const response = {
  me: {
    user: {
      id: 1,
      firstname: "Admin",
      lastname: "User",
      fullname: "User fullname",
      email: "admin@totara.com"
    },
    system: {
      wwwroot: "http://10.0.8.127/",
      apiurl: "http://10.0.8.127/totara/mobile/api.php",
      release: "Evergreen (Build: 20190619.00)",
      request_policy_agreement: false,
      request_user_consent: false,
      request_user_fields: true
    },
    enrolled_courses: {
      id: 1,
      fullname: "User",
      shortname: "Usr",
      idnumber: 1
    }
  }
};

const mocks = [
  {
    request: {
      query: QueryMe
    },
    result: {
      data: response
    }
  }
];

const mocksError = [
  {
    request: {
      query: QueryMe
    },
    error: new Error("Error")
  }
];

describe("AdditionalActionRule, Apollo MockedProvider should test three state such as loading, final and error ", () => {
  it("Test result : Render loading state initially and return no component", () => {
    const component = renderer.create(
      <MockedProvider mocks={[]} addTypename={false}>
        <AdditionalActionRule />
      </MockedProvider>
    );
    const tree = component.toJSON();
    expect(tree).toBeNull();
  });

  it("Test result : Once render additional action rule, it will return child component", async () => {
    const component = renderer.create(
      <MockedProvider mocks={mocks} addTypename={false}>
        <AdditionalActionRule>
          <Text>Modal Details</Text>
        </AdditionalActionRule>
      </MockedProvider>
    );

    await act(async () => {
      await wait(0);
    });
    const tree = component.root.props.children.props.children.props.children;
    expect(tree).toContain("Modal Details");
  });

  it("Test result : Set mock-error response and return Error UI", async () => {
    const component = renderer.create(
      <MockedProvider mocks={mocksError} addTypename={false}>
        <AdditionalActionRule />
      </MockedProvider>
    );
    await act(async () => {
      await wait(0);
    });
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe("AdditionalActionRuleCondition, Passing different bool params which relate policyAgreement,userConsent,userFields and get a bool value", () => {
  it("It should return true for any of passing booleans are true, and return false for all the passing values are false", () => {
    let actualResultSingle = AdditionalActionRuleCondition(
      true,
      false,
      false,
      false
    );
    expect(actualResultSingle).toBe(true);

    actualResultSingle = AdditionalActionRuleCondition(
      false,
      true,
      false,
      false
    );
    expect(actualResultSingle).toBe(true);

    actualResultSingle = AdditionalActionRuleCondition(
      false,
      false,
      true,
      false
    );
    expect(actualResultSingle).toBe(true);

    actualResultSingle = AdditionalActionRuleCondition(
      false,
      false,
      false,
      true
    );
    expect(actualResultSingle).toBe(true);

    let actualResultDouble = AdditionalActionRuleCondition(
      true,
      true,
      false,
      false
    );
    expect(actualResultDouble).toBe(true);

    actualResultDouble = AdditionalActionRuleCondition(
      true,
      false,
      true,
      false
    );
    expect(actualResultDouble).toBe(true);

    actualResultDouble = AdditionalActionRuleCondition(
      true,
      false,
      false,
      true
    );
    expect(actualResultDouble).toBe(true);

    actualResultDouble = AdditionalActionRuleCondition(
      false,
      true,
      true,
      false
    );
    expect(actualResultDouble).toBe(true);

    actualResultDouble = AdditionalActionRuleCondition(
      false,
      true,
      false,
      true
    );
    expect(actualResultDouble).toBe(true);

    actualResultDouble = AdditionalActionRuleCondition(
      false,
      false,
      true,
      true
    );
    expect(actualResultDouble).toBe(true);

    let actualResultThribble = AdditionalActionRuleCondition(
      true,
      true,
      true,
      false
    );
    expect(actualResultThribble).toBe(true);

    actualResultThribble = AdditionalActionRuleCondition(
      true,
      true,
      false,
      true
    );
    expect(actualResultThribble).toBe(true);

    actualResultThribble = AdditionalActionRuleCondition(
      true,
      false,
      true,
      true
    );
    expect(actualResultThribble).toBe(true);

    actualResultThribble = AdditionalActionRuleCondition(
      false,
      true,
      true,
      true
    );
    expect(actualResultThribble).toBe(true);

    const actualResultAll = AdditionalActionRuleCondition(
      true,
      true,
      true,
      true
    );
    expect(actualResultAll).toBe(true);

    const actualResultNone = AdditionalActionRuleCondition(
      false,
      false,
      false,
      false
    );
    expect(actualResultNone).toBe(false);
  });
});
