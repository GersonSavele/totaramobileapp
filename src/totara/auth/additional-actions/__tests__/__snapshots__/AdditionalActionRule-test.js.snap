// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AdditionalActionRule, Apollo MockedProvider should test three state such as loading, final and error  Test result : Set mock-error response and return Error UI 1`] = `
<Modal
  hardwareAccelerated={false}
  visible={true}
>
  <RCTSafeAreaView
    emulateUnlessSupported={true}
    style={
      Object {
        "backgroundColor": "rgba(0, 0, 0, 0.7)",
        "bottom": 0,
        "flex": 1,
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
  >
    <View
      style={
        Object {
          "backgroundColor": "#FFF",
          "borderRadius": 4,
          "bottom": 32,
          "flex": 1,
          "left": 20,
          "padding": 5,
          "position": "absolute",
          "right": 20,
          "top": 32,
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "flex": 1,
            "flexDirection": "column",
            "justifyContent": "center",
            "marginLeft": "5%",
            "marginRight": "5%",
          }
        }
      >
        <View
          style={
            Object {
              "alignItems": "center",
            }
          }
        >
          <View
            style={
              Object {
                "alignItems": "center",
                "marginBottom": 16,
                "marginTop": 16,
              }
            }
          >
            <Image
              source={
                Object {
                  "testUri": "../../../src/resources/images/general_error/general_error.png",
                }
              }
              style={
                Object {
                  "alignItems": "center",
                  "height": 375,
                  "resizeMode": "contain",
                  "width": 525,
                }
              }
            />
          </View>
          <View
            style={
              Object {
                "alignItems": "center",
                "marginBottom": 16,
                "marginTop": 16,
              }
            }
          >
            <View
              style={
                Object {
                  "alignItems": "flex-start",
                  "paddingBottom": 16,
                }
              }
            >
              <Text
                style={
                  Array [
                    Object {
                      "textAlign": "center",
                    },
                    Object {
                      "color": "#3D444B",
                      "fontSize": 44,
                      "fontWeight": "600",
                    },
                  ]
                }
              >
                translated general_error_feedback-modal.title with undefined
              </Text>
            </View>
            <View
              style={
                Object {
                  "alignItems": "flex-start",
                  "paddingBottom": 16,
                }
              }
            >
              <Text
                style={
                  Array [
                    Object {
                      "textAlign": "center",
                    },
                    Object {
                      "color": "#3D444B",
                      "fontSize": 29,
                      "fontWeight": "100",
                    },
                  ]
                }
              >
                translated general_error_feedback-modal.description with undefined
              </Text>
            </View>
          </View>
          <View
            accessibilityRole="button"
            accessibilityStates={Array []}
            accessible={true}
            isTVSelectable={true}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "opacity": 1,
              }
            }
          >
            <View
              style={
                Array [
                  Object {},
                ]
              }
            >
              <Text
                style={
                  Array [
                    Object {
                      "color": "#007AFF",
                      "fontSize": 18,
                      "padding": 8,
                      "textAlign": "center",
                    },
                  ]
                }
              >
                this is a mock button
              </Text>
            </View>
          </View>
          <View
            accessibilityRole="button"
            accessibilityStates={Array []}
            accessible={true}
            isTVSelectable={true}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "opacity": 1,
              }
            }
          >
            <View
              style={
                Array [
                  Object {},
                ]
              }
            >
              <Text
                style={
                  Array [
                    Object {
                      "color": "#007AFF",
                      "fontSize": 18,
                      "padding": 8,
                      "textAlign": "center",
                    },
                  ]
                }
              >
                this is a mock button
              </Text>
            </View>
          </View>
        </View>
      </View>
    </View>
  </RCTSafeAreaView>
</Modal>
`;
